cql3/statements/select_statement.cc

```diff
@@ -1154,49 +1156,82 @@ indexed_table_select_statement::do_execute(query_processor& qp,
 
                 if (whole_partitions || partition_slices) {
                     tracing::trace(state.get_trace_state(), "Consulting index {} for a single slice of keys, aggregation query", _index.metadata().name());
-                    return find_index_partition_ranges(qp, state, *internal_options).then(utils::result_wrap_unpack(
-                            [this, now, &state, &internal_options, &qp, consume_results = std::move(consume_results)] (dht::partition_range_vector partition_ranges, lw_shared_ptr<const service::pager::paging_state> paging_state) {
-                        return do_execute_base_query(qp, std::move(partition_ranges), state, *internal_options, now, paging_state)
-                        .then(utils::result_wrap_unpack([paging_state, consume_results = std::move(consume_results)](foreign_ptr<lw_shared_ptr<query::result>> results, lw_shared_ptr<query::read_command> cmd) {
-                            return consume_results(std::move(results), std::move(cmd), std::move(paging_state));
-                        }));
-                    }));
+                    auto result_partition_ranges_and_paging_state = co_await find_index_partition_ranges(qp, state, *internal_options);
+                    if (result_partition_ranges_and_paging_state.has_error()) {
+                        co_return failed_result_to_result_message(std::move(result_partition_ranges_and_paging_state));
+                    }
```

Comment:
This looks a bit like go
```go
if err != nil {
    return err;
}
```

I feel like there should be a better way to do this. Maybe by using exceptions on failure (but with `coroutine::return_exception` so we don't throw) and then converting them to result message somewhere high in the call stack.

Reply:
> This looks a bit like go
> 
> ```go
> if err != nil {
>     return err;
> }
> ```

That's exactly what it is.

> 
> I feel like there should be a better way to do this. Maybe by using exceptions on failure (but with `coroutine::return_exception` so we don't throw) and then converting them to result message somewhere high in the call stack.

The whole idea here is to avoid exceptions in the data path, since they are extremely expensive. See https://www.scylladb.com/presentations/retaining-goodput-with-query-rate-limiting/ from @piodul.

There's talks in the C++ community to have more flexible control flow, but nothing firm.